services/src/main/java/org/keycloak/userprofile/utils/StoredUserProfile.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.userprofile.utils;
+
+import org.keycloak.userprofile.UserProfile;
+
+import java.util.List;
+
+/**
+ * Abstraction, which allows to update the user in various contexts (Required action of already existing user, or first identity provider
+ * login when user doesn't yet exists in Keycloak DB)
+ *
+ * @author <a href="mailto:markus.till@bosch.io">Markus Till</a>
+ */
+public interface StoredUserProfile extends UserProfile {
```

Comment:
I understand the reason behind it I think. So that a `UserProfile` is marked as writable.

After reading through your code, looks like to me that the best would be if we had a single `UserProfile` implementation, with attributes encapsulated via a `Attributes` API so that the read/write capability could be controlled accordingly depending on the type of the event.

Reply:
I'm a bit confused to be honest. the current UserProfile interface is actually very close to an Attributes interface. So to move towards your solution I would rename UserProfile to UserAttributes and merge the StoredUserProfile methods into this interface. Then on all implementations I would add the write methods and throw Exceptions when they are called.

Is this even close to your idea? I will do so. But may I got you fully wrong here, I at least suspect that.